rfcs/0000-sudt-info/0000-sudt info.md

```diff
@@ -0,0 +1,184 @@
+---
+Number: "0000"
+Category: Standards Track
+Status: Proposal
+Author: Xiang Cheng
+Organization: Lay2dev
+Created: 2021-12-25
+---
+
+# sudt info
+
+## 1.Overview
+
+sudt is a very simple udt standard. With the udt deployed by this standard, we can only get the balance from the chain, but not the token name, token symbol, decimals and other parameters. 
+
+To extend the existing sudt standard, it is necessary to design a scheme to store token information and bind it to the corresponding sudt, but does not destroy any compatibility with the existing sudt.
+
+## 2. sudt Info format
+
+In sudt info, the most basic information that must be included is token_name, token_symbol, and decimals. As for total supply and balance, they can all be obtained by indexing the cell.
+
+A sudt Info in specification looks like following:
+
+```
+decimals: uint8
+name: Common knowledge Byte
+symbol: CKB
+extra: {website:"https://nervos.org"} // optional
+```
+
+The serialization format of sudt info should be json. 
```

Comment:
I'm curious if Molecule can be used here as it's more space-efficient and a default on CKB. If I remember correctly, the reason for using JSON than Molecule was the ease of use for application developers? Not sure if [Molecule's javascript support](https://github.com/nervosnetwork/moleculec-es) renders the reason obsolete?

Reply:
@xcshuan any thoughts on this one?

cc @CipherWang @louzhixian

Reply:
Maybe we can use the simplest format, which is length+content.

Reply:
A simple format is consistent with sUDT indeed.

Reply:
> with the drawback of lacking deserialization tools.

This simple encoding scheme should be very easy to deserialize?

Reply:
We're considering 3 options:
1. JSON - easy to use, space inefficient;
2. Molecule - not that easy, space efficient, the default serialization format used by CKB (like RLP in Ethereum);
3. length + content simple encoding - easy or not? More space efficient than JSON.

My opinion is to go with either 1 or 2. Option 3 is basically an ad-hoc version of 2 and will create standard fragmentation.   Personally I think molecule is not that hard (especially when compare with 3), and could become easier to use if we gather our efforts to create better tools. JSON has a big advantage in this case because SUDT Info will mostly be used by dapps, not on-chain scripts.

Reply:
I think we have to think about the user cases here. 
Are these info just parsed off-chain or they are gonna be used in on-chain contracts?

The trade-off here except for what Jan mentions:
- JSON is hard to parse on-chain, but easy to parse off-chain.
- For the common fields like name, symbol and decimal, molecule is good. But for the `extra` field, it needs additional schema to describe the inner key of `extra` field, which is very inconvenient in most scenarios.

Since the sudt info is supplemental proposal, I think an on-chain contract can barely depends  on this to handle any logics. Then JSON is good enough.

## space difference
- molecule 
  - 4(total length) + 4 (offset) * 4 = 20 bytes except for the values themselves
- JSON
  - `{"decimals":16, "name":"","symbol":"","extra":""}` takes 47 ~ 49 bytes(`uint8` in JSON can be 1 ~ 3 bytes) except for the values themselves.

```
decimals: uint8
name: Common knowledge Byte
symbol: CKB
extra: {website:"https://nervos.org"} // optional
```

The difference is ~28 CKBytes. I think it is acceptable to use JSON.

Reply:
Thanks @Keith-CY @huwenchao for the inputs!

I agree that 28 CKBytes is an acceptable trade-off given only 1 SUDT info cell is needed for a SUDT. Maybe we could stick with the original idea of using JSON here. @xcshuan thoughts?

Reply:
> Maybe we could stick with the original idea of using JSON here.

At least we need to verify in the script whether the data of the info cell is a valid SUDT Info. I don't know if using json will make this troublesome.

Reply:
In this case, they are almost the same in size.

What if we introduce another token info scheme that consists of various fields, e.g. a token scheme of NFT/FT under the drafting stage has the following keys:
  - version: identify the scheme version
  - name:
  - symbol:
  - description:
  - image:
  - audio:
  - video:
  - model:
  - scheme: scheme of the `characteristic` field in MiBao NFT cell
  - properties: arbitrary user-defined data for UI
  - localization: identify how to render an NFT by locale info

The `properties` value could be oversize and Molecule would be preferable in this case.

If SUDT picks JSON while other project/token picks Molecule, dapp developers have to keep them in mind.

Reply:
If there is a library for molecule available, I think it should be better to use molecule.

Reply:
> If there is a library for molecule available, I think it should be better to use molecule.

There are libraries for C, Rust, go and JavaScript: <https://github.com/nervosnetwork/molecule>.

> At least we need to verify in the script whether the data of the info cell is a valid SUDT Info. I don't know if using json will make this troublesome.

It will. Use JSON in on-chain contract is hard, expensive and may bring security risks as Jindong mentions. Do we really need to verify sudt info format in scripts? It doesn't make sense in my opinion.

If we skip the verification, we can stick to the original method.

If we **do** need that, I recommend use the format below:

```
vector Bytes <byte>;

table SudtInfo {
    decimals: byte,
    name: Bytes,
    symbol: Bytes,
    extra: Bytes,
}
```

The `extra` may be a problem in real world cases. Since molecule needs schema to parse data, but different user may  use different fields here. The best solution for this field I can find is to use an embedded JSON string. In this case, on-chain contracts will only verify the fixed fields with molecule, and we have to develop some tools for users in the SDK to parse the whole struct.

Reply:
I prefer not to verify format in CKB script. It simplifies the script/standard a lot. If someone creates an ill-formated SUDT info cell intentionally or accidentally, applications can simply ignore the bad info cell and the result is the same as no SUDT info cell.

Reply:
Summary: we're considering JSON vs. Molecule, while JSON is simple and widely used, its drawback is more difficult to parse/validate on chain. However it seems unnecessary to include SUDT info parse/validation in on-chain script, which basically avoids the problem. In this case I suggest to go with JSON and skip on-chain format validation, to keep things simple.

Reply:
@xcshuan I suggest adding @huwenchao as 2nd author because it's his analysis and suggestions and having him complete the RFC.

Reply:
> adding @huwenchao as 2nd author

Ok, I think it's fine.